carball/analysis/events/carry_detection.py

```diff
@@ -0,0 +1,187 @@
+            else:
+                if previous_start == -1:
+                    previous_start = start_frames[frame_index]
+                previous_end = end_frames[frame_index + 1]
+
+        if previous_start != -1:
+            merged_start.append(previous_start)
+            merged_end.append(previous_end)
+        player_carry_data.start_frames = merged_start
+        player_carry_data.end_frames = merged_end
+
+    def expand_carries(self, carry_data: CarryData, player_carry_data: CarryData,
+                       player: Player, proto_game: game_pb2.Game) -> CarryData:
+        """
+        Expands / merges carries based on hit list.
+        If nothing is modified then this returns valid_frames
+        """
+
+        self.merge_dribbles(carry_data, player_carry_data)
+        start_frames = player_carry_data.start_frames
+        end_frames = player_carry_data.end_frames
+
+        starting_hit_index = 0
+        hit_list = proto_game.game_stats.hits
+
+        # to expand you look at all hits before the start frame and check if they are valid for this purpose.
+
+
+        # Need to start cutting off dribbles if someone other than the player hits the ball and the player does not gain another hit on the ball before the dribble ends.
+
+        """
+        for frame_index in range(len(start_frames)):
+            for i in range(starting_hit_index, len(hit_list)):
+                if hit_list[i].frame_number < start_frames[frame_index]:
+                    continue
+                if hit_list[i].frame_number > start_frames[frame_index]:
+                    # now we go backwards
+                    hit_number = i
+                    if hit_list[i].frame_number > end_frames[frame_index]:
+                        hit_number -= 1
+                    while (hit_number > 0 and hit_list[hit_number].player_id.id == player.id.id and
+                           hit_list[hit_number - 1].frame_number in valid_frames.index):
+                        hit_number -= 1
+                    # now we have the first hit that is in valid
+                    if hit_list[hit_number].frame_number < start_frames[frame_index]:
+                        # now we modify
+                        print('NEED MODIFICATION')
+
+            if starting_hit_index >= len(hit_list) - 1:
+                logger.warning("We have reached a point where dribbles should not exist anymore")
+        # Look for the first carry frame and find all hits before it by the same player
+        pass
+        """
+
+        return player_carry_data
+
+    def get_distance(self, player_frames: pd.DataFrame, start: int, end: int) -> pd.DataFrame:
+        x1 = player_frames.pos_x[start]
+        y1 = player_frames.pos_y[start]
+
+        x2 = player_frames.pos_x[end]
+        y2 = player_frames.pos_y[end]
+
+        return ((x2 - x1) ** 2 + (y2 - y1) ** 2) ** 0.5
+
+    def creat_start_end_frames(self, carry_frames: pd.DataFrame) -> Tuple[List[int], List[int]]:
```

Comment:
Apart from the typo in the function name, what're you trying to do here w/ start frames and end_frames? It feels like there has to be a better way.

Reply:
I am trying to get the first frame of the dribble and the last frame of the dribble which is then very useful later on

Reply:
There's a bit of a random threshold of 3.

Reply:
Because i have seen replays where the frame number is skipped by 1 or 2.